### PR TITLE
[opt](Nereids) support where, group by, having, order by clause without from clause in query statement

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Support where, group by, having, order by clause without from clause in query statement.
**For example as following:**

SELECT 1 AS a, COUNT(), SUM(2), AVG(1), RANK() OVER() AS w_rank
WHERE 1 = 1
GROUP BY a, w_rank
HAVING COUNT() IN (1, 2) AND w_rank = 1
ORDER BY a;

this will return result:

| a  |count(*)|sum(2)|avg(1)|w_rank|
+----+--------+------+------+------+
| 1  |       1|     2|   1.0|     1|

**For another example as following:**
select 1 c1, 2 union (select "hell0", "") order by c1
the second column datatype will be varchar(65533), 65533 is the default varchar length.
this will return result:

|c1    | 2 |
+------+---+
|1     | 2 |
|hell0 |   |

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

File: fe/fe-common/src/main/java/org/apache/doris/catalog/ScalarType.java [hunk 1]
```diff
@@ -1060,7 +1060,13 @@ public static ScalarType getAssignmentCompatibleType(
             if (t1.type == PrimitiveType.STRING || t2.type == PrimitiveType.STRING) {
                 return createStringType();
             }
-            return createVarcharType(Math.max(t1.len, t2.len));
+            int minLength = Math.min(t1.len, t2.len);
+            if (minLength < 0) {
+                // If < 0 which means max length, use firstly
+                return createVarcharType(minLength);
+            }
```
Comment (XieJiann) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/ScalarType.java:1067]:
```suggestion
            if (t1.len < 0 ｜｜ t2.len < 0) {
                // If < 0 which means max length, use firstly
                return createVarcharType(minLength);
            }
```

**Review by keanji-x — APPROVED** [review]

**Pull request merged**